Sources/StreamChat/Repositories/SyncRepository.swift

```diff
@@ -0,0 +1,424 @@
+                    done(.continue)
+                    return
+                }
+
+                self?.syncMissingEvents(
+                    using: lastPendingConnectionDate,
+                    channelIds: preSyncChannelIds,
+                    bumpLastSync: false
+                ) { result in
+                    switch result {
+                    case let .success(channelIds):
+                        synchedChannelIds = Set(channelIds)
+                        // As per our sync logic, we should keep the last connection date.
+                        self?.updateUserValue(
+                            { $0?.lastPendingConnectionDate = self?.lastConnection },
+                            completion: { _ in done(.continue) }
+                        )
+                    case let .failure(error):
+                        synchedChannelIds = Set([])
+                        done(error.shouldRetry ? .retry : .continue)
+                    }
+                }
+            }
+        }
+        operations.append(syncEvents)
+
+        // 2. Start watching open channels.
+        // We keep track of the watched channels to avoid clearing them out in future steps
+        var watchedChannelIds: Set<ChannelId> = []
+        let refetchChannelOperations: [AsyncOperation] = activeChannelControllers.allObjects.compactMap { controller in
+            // Reset only the controllers that need recovery
+            guard controller.isAvailableOnRemote else { return nil }
+            if let cid = controller.cid, synchedChannelIds.contains(cid) {
+                return nil
+            }
+            return AsyncOperation(maxRetries: maxRetriesCount) { [weak self] done in
+                log.info("2. Start watching open channel", subsystems: .offlineSupport)
+                self?.watchActiveChannel(for: controller) { result in
+                    switch result {
+                    case let .success(channelId):
+                        watchedChannelIds.insert(channelId)
+                        done(.continue)
+                    case let .failure(error):
+                        done(error.shouldRetry ? .retry : .continue)
+                    }
+                }
+            }
+        }
+        operations.append(contentsOf: refetchChannelOperations)
+
+        // 3. Refetch channel lists queries, link only what backend returns (the 1st page)
+        // 4. Clean up local message history for channels that are outdated/will get outdated
+        // We use `synchedChannelIds` to keep track of the channels that were synched both in the previous step and
+        // after each ChannelListController recovery.
+        let refetchChannelListQueryOperations: [AsyncOperation] = activeChannelListControllers.allObjects.compactMap { controller in
+            // Reset only the controllers that need recovery
+            guard controller.isAvailableOnRemote else { return nil }
```

Comment:
Nitpick
Something like `.filter(\.isAvailableOnRemote)` would be preferrable over here